### PR TITLE
make compile_templates deterministic

### LINKED ISSUE #1452: `compile_templates` is non-deterministic on `{% import ... as ... with context %}`
Following code generates non-deterministic compiled output.
```python
from jinja2 import loaders
from jinja2.environment import Environment

env = Environment(loader=loaders.DictLoader({"foo": """
{% import "bar" as bar with context %}
{% import "baz" as baz with context %}
"""}))

env.compile_templates('archive', zip=None, ignore_errors=False)
```

```diff
 from __future__ import generator_stop
 from jinja2.runtime import LoopContext, Macro, Markup, Namespace, TemplateNotFound, TemplateReference, TemplateRuntimeError, Undefined, concat, escape, identity, internalcode, markup_join, missing, str_join
 name = 'foo'

 def root(context, missing=missing):
     resolve = context.resolve_or_missing
     undefined = environment.undefined
     cond_expr_undefined = Undefined
     if 0: yield None
     l_0_bar = l_0_baz = missing
     pass
     yield '\n'
-    l_0_bar = context.vars['bar'] = environment.get_template('bar', 'foo').make_module(context.get_all(), True, {'bar': l_0_bar, 'baz': l_0_baz})
+    l_0_bar = context.vars['bar'] = environment.get_template('bar', 'foo').make_module(context.get_all(), True, {'baz': l_0_baz, 'bar': l_0_bar})
     context.exported_vars.discard('bar')
     yield '\n'
-    l_0_baz = context.vars['baz'] = environment.get_template('baz', 'foo').make_module(context.get_all(), True, {'bar': l_0_bar, 'baz': l_0_baz})
+    l_0_baz = context.vars['baz'] = environment.get_template('baz', 'foo').make_module(context.get_all(), True, {'baz': l_0_baz, 'bar': l_0_bar})
     context.exported_vars.discard('baz')

 blocks = {}
 debug_info = '2=13&3=16'
```

Chromium uses jinja2 for code generator during build, and this breaks [Reproducible builds](https://en.wikipedia.org/wiki/Reproducible_builds) when we are switching to python3.

Some more context is in https://crbug.com/1194274

Environment:

- Python version: `Python 3.9.2`
- Jinja version: `3.0.1`

### LINKED ISSUE #1453: compile_templates is non-deterministic when using filters
Following code generates non-deterministic compiled output.

```python
from jinja2 import loaders
from jinja2.environment import Environment

env = Environment(loader=loaders.DictLoader({"foo": """
{{ 1|filter1 }}
{{ 2|filter2 }}
"""}))

env.filters["filter1"]=lambda value: value
env.filters["filter2"]=lambda value: value

env.compile_templates('archive', zip=None, ignore_errors=False)
```

```diff
 from __future__ import generator_stop
 from jinja2.runtime import LoopContext, Macro, Markup, Namespace, TemplateNotFound, TemplateReference, TemplateRuntimeError, Undefined, concat, escape, identity, internalcode, markup_join, missing, str_join
 name = 'foo'

 def root(context, missing=missing):
     resolve = context.resolve_or_missing
     undefined = environment.undefined
     cond_expr_undefined = Undefined
     if 0: yield None
     try:
-        t_1 = environment.filters['filter1']
+        t_1 = environment.filters['filter2']
     except KeyError:
         @internalcode
         def t_1(*unused):
-            raise TemplateRuntimeError("No filter named 'filter1' found.")
+            raise TemplateRuntimeError("No filter named 'filter2' found.")
     try:
-        t_2 = environment.filters['filter2']
+        t_2 = environment.filters['filter1']
     except KeyError:
         @internalcode
         def t_2(*unused):
-            raise TemplateRuntimeError("No filter named 'filter2' found.")
+            raise TemplateRuntimeError("No filter named 'filter1' found.")
     pass
     yield '\n1\n2'

 blocks = {}
 debug_info = ''
```

Chromium uses jinja2 for code generator during build, and this breaks [Reproducible builds](https://en.wikipedia.org/wiki/Reproducible_builds) when we are switching to python3.

Some more context is in https://crbug.com/1194274

Environment:

- Python version: `Python 3.9.2`
- Jinja version: `3.0.1`

### PR DESCRIPTION
**Pull request by atetubou**
Python3 doesn't keep insertion order for set(), so this sorts some
places for deterministic output for compiled template.

- fixes #1452
- fixes #1453

Checklist:

- [x] Add tests that demonstrate the correct behavior of the change. Tests should fail without the change.
- [ ] Add or update relevant docs, in the docs folder and in code.
- [x] Add an entry in `CHANGES.rst` summarizing the change and linking to the issue.
- [ ] Add `.. versionchanged::` entries in any relevant code docs.
- [x] Run `pre-commit` hooks and fix any issues.
- [ ] Run `pytest` and `tox`, no tests failed.

### CODE REVIEW

**Comment by davidism** [comment]
Please rebase to 3.0.x instead of main.

File: tests/test_compile.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        env.filters["filter2"] = lambda value: value
+
+        env.compile_templates(self.archive, zip=None, ignore_errors=False)
+
+        files = os.listdir(self.archive)
+        assert len(files) == 1
+
+        with open(os.path.join(self.archive, files[0])) as f:
+            assert f.read().splitlines() == [
+                "from __future__ import generator_stop",
```
Comment (davidism) [inline tests/test_compile.py:40]: These tests will be very brittle, they'll need to be updated every time something unrelated changes the compiled code. Ideally, determine a way to test this without inspecting the generated code, or at least only look for a specific pattern.

File: CHANGES.rst [hunk 2]
```diff
@@ -11,6 +11,7 @@ Version 3.0.2
 
 Unreleased
 
+- Make ``compile_templates`` deterministic. :issue:`1452, 1453`
```
Comment (davidism) [inline CHANGES.rst:14]: Two empty lines between sections, three spaces after bullet.

File: tests/test_compile.py (hunk 1 above)
Comment (atetubou) [reply tests/test_compile.py]: Yeah, I changed to check appearence order of some keywords.

**Comment by atetubou** [comment]
Thank you for review, I think I addressed your comments.

**Comment by atetubou** [comment]
Can I have a review again?

**Comment by atetubou** [comment]
ping?

**Comment by davidism** [comment]
Please don't bump posts like that. Add volunteers, we will get to your pr when we have time available.

**Comment by atetubou** [comment]
Sorry for pinging and thank you for your merge!
